### ISSUE TITLE
Running inference with a custom YOLOv3 model

### ISSUE THREAD

**Issue by jhara23** [openvinotoolkit/dlstreamer_gst#189]
Hi, I trained a custom YOLOv3, and converted the model to IR successfully.  
Then, I tried to run inference using gvadetect, but it didn't detect anything.
Could you help identify what's wrong with my settings?  

I tried the same steps on the custom YOLOv3 model provided at the following site, but the result was the same.  
[https://github.com/cansik/yolo-hand-detection/releases/download/pretrained/cross-hands-tiny.cfg](https://github.com/cansik/yolo-hand-detection/releases/download/pretrained/cross-hands-tiny.cfg)  
[https://github.com/cansik/yolo-hand-detection/releases/download/pretrained/cross-hands-tiny.weights](https://github.com/cansik/yolo-hand-detection/releases/download/pretrained/cross-hands-tiny.weights)
I tried on the normal YOLOv3 version of above models, but the result was same as well.  

The commands and files I used for the above model is as below.  
I'm using OpenVINO version 2021.2.185 on Ubuntu 18.04.  

1. converting YOLOv3 to Tensorflow

```bash
python3 tensorflow-yolo-v3/convert_weights_pb.py --class_names cross-hands-tiny.names --data_format NHWC --weights_file cross-hands-tiny.weights --tiny
```

cross-hands-tiny.names

```txt
hand
```

2. converting Tensorflow to IR

```bash
python3 /opt/intel/openvino_2021/deployment_tools/model_optimizer/mo_tf.py --input_model frozen_darknet_yolov3_model.pb --transformations_config cross-hands-tiny.json --batch 1
```

cross-hands-tiny.json - only changed "classes" from /opt/intel/openvino_2021/deployment_tools/model_optimizer/extensions/front/tf/yolo_v3_tiny.json

```json
[
  {
    "id": "TFYOLOV3",
    "match_kind": "general",
    "custom_attributes": {
      "classes": 1,
      "anchors": [10, 14, 23, 27, 37, 58, 81, 82, 135, 169, 344, 319],
      "coords": 4,
      "num": 6,
      "masks": [[3, 4, 5], [0, 1, 2]],
      "entry_points": ["detector/yolo-v3-tiny/Reshape", "detector/yolo-v3-tiny/Reshape_4"]
    }
  }
]
```

3. run inference with gvadetect

```bash
gst-launch-1.0 urisourcebin buffer-size=4096 uri=https://raw.githubusercontent.com/cansik/yolo-hand-detection/master/images/zachary-nelson-98Elr-LIvD8-unsplash.jpg \
    ! decodebin ! imagefreeze \
    ! gvadetect model=frozen_darknet_yolov3_model.xml threshold=0.5 model-proc=cross-hands-tiny-model-proc.json \
    ! gvawatermark \
    ! autovideosink
```

cross-hands-tiny-model-proc.json - only changed "classes" and "labels" from [https://raw.githubusercontent.com/openvinotoolkit/dlstreamer_gst/master/samples/model_proc/yolo-v3-tiny.json](https://raw.githubusercontent.com/openvinotoolkit/dlstreamer_gst/master/samples/model_proc/yolo-v3-tiny.json)

```json
{
    "json_schema_version": "2.0.0",
    "input_preproc": [],
    "output_postproc": [
        {
            "converter": "tensor_to_bbox_yolo_v3",
            "iou_threshold": 0.5,
            "classes": 1,
            "anchors": [
                10.0,
                14.0,
                23.0,
                27.0,
                37.0,
                58.0,
                81.0,
                82.0,
                135.0,
                169.0,
                344.0,
                319.0
            ],
            "masks": [
                3,
                4,
                5,
                1,
                2,
                3
            ],
            "bbox_number_on_cell": 3,
            "cells_number": 13,
            "labels": [
                "hand"
            ]
        }
    ]
}
```

Best,  
Junichi

**Comment by jhara23** [comment]
It seems I found the solution.  
When converting to Tensorflow to IR, "--reverse_input_channels" option is needed.  
I'm not sure why it's needed, though.  

2. converting Tensorflow to IR

```bash
python3 /opt/intel/openvino_2021/deployment_tools/model_optimizer/mo_tf.py --input_model frozen_darknet_yolov3_model.pb --transformations_config cross-hands-tiny.json --batch 1 --reverse_input_channels
```

Best,
Junichi

**Comment by mliu2020** [comment]
Hi @jhara23,

If my understanding is correct, your application is to detect a hand from the input video, so you are expecting an output video with a bound box around the hand at the output video.

As you might understand, DL Streamer acts as an integration framework for the OpenVINO inference engine, it makes it easier for video to be processed by IE, so whatever works in DL Streamer, it should work with OpenVINO application.

To your case, it seems work now and for the "--reverse_input_channels" option, you should check what is the input color channel ordering of the YoloV3 from its input layer. Then you should find out what is the pipeline input color channel order to see if it was swapped. One thing you can see is the input color format of [gvadetect ](https://github.com/openvinotoolkit/dlstreamer_gst/wiki/gvadetect)is BGRx.

Mark Liu

**Comment by jhara23** [comment]
Hi @mliu2020,

Thank you so much for your comment!
I should have tried to run inference with IE to see if the conversion was correct...
But I'm prefer using DL Streamer only since it's much easier and quicker than using IE.

I don't have any further questions now, so let me close this issue.

Thanks,
Junichi
